1.8.0 Gradle changes (#3333)
JetBrains/kotlin-web-site: docs/topics/gradle/gradle-plugin-variants.md

Learn more about [variant selection in Gradle](https://docs.gradle.org/current/userguide/variant_model.html).

With Gradle plugin variants, the Kotlin team can ship different Kotlin Gradle plugin (KGP) variants for different Gradle versions. 
The goal is to support the base Kotlin compilation in the `main` variant, which corresponds to the oldest supported versions of
Gradle. Each variant will have implementations for Gradle features from a corresponding release. The latest variant will
support the latest Gradle feature set. With this approach, it is possible to extend support for older Gradle versions 
with limited functionality.

Currently, there are the following variants of the Kotlin Gradle plugin:

| Variant's name | Corresponding Gradle versions |
|----------------|-------------------------------|
| `main`         | 6.8.3–6.9.2                   |
| `gradle70`     | 7.0                           |
| `gradle71`     | 7.1-7.4                       |
| `gradle75`     | 7.5                           |
| `gradle76`     | 7.6 and higher                |

> In future Kotlin releases, more variants will be probably added.

Reviewer: ```suggestion
In future Kotlin releases, more variants will probably be added.
```

To check which variant your build uses, enable
the [`--info` log level](https://docs.gradle.org/current/userguide/logging.html#sec:choosing_a_log_level) and find a
string in the output starting with `Using Kotlin Gradle plugin`, for example, `Using Kotlin Gradle plugin main variant`.

## Troubleshooting

> Here are workarounds for some known issues with variant selection in Gradle:
> * [ResolutionStrategy in pluginManagement is not working for plugins with multivariants](https://github.com/gradle/gradle/issues/20545)
> * [Plugin variants are ignored when a plugin is added as the `buildSrc` common dependency](https://github.com/gradle/gradle/issues/20847)
>
{type="note"}

### Gradle can't select a KGP variant in a custom configuration

This is an expected situation that Gradle can't select a KGP variant in a custom configuration.
If you use a custom Gradle configuration:

<tabs group="build-script">
<tab title="Kotlin" group-key="kotlin">

```kotlin
configurations.register("customConfiguraton") {
    ...
}
```

</tab>
<tab title="Groovy" group-key="groovy">

```groovy
configurations.register("customConfiguraton") {
    ...
}
```

</tab>
</tabs>

> and want to add a dependency on the Kotlin Gradle plugin, for example:

Reviewer: ```suggestion
And want to add a dependency on the Kotlin Gradle plugin, for example:
```

[//]: # (title: Support for Gradle plugin variants)

Gradle 7.0 introduced a new feature for Gradle plugin authors
— [plugins with variants](https://docs.gradle.org/7.0/userguide/implementing_gradle_plugins.html#plugin-with-variants).

> This feature makes it easier to add support for new Gradle features while maintaining compatibility for Gradle versions

Reviewer 2: Better imho:
> This feature makes it easier to add support for lastest Gradle features while maintaining compatibility for older Gradle versions

> This feature makes it easier to add support for new Gradle features while maintaining compatibility for Gradle versions

Author: ```suggestion
This feature makes it easier to add support for latest Gradle features while maintaining compatibility with older Gradle versions
```

below 7.1. Learn more about [variant selection in Gradle](https://docs.gradle.org/current/userguide/variant_model.html).

With Gradle plugin variants, the Kotlin team can ship different Kotlin Gradle plugin (KGP) variants for different Gradle versions. 
The goal is to support the base Kotlin compilation in the `main` variant, which corresponds to the oldest supported versions of
Gradle. Each variant will have implementations for Gradle features from a corresponding release. The latest variant will

> support the widest Gradle feature set. With this approach, it is possible to extend support for older Gradle versions 

Reviewer 2: Not sure about "widest" - maybe better to use "latest"?

> support the widest Gradle feature set. With this approach, it is possible to extend support for older Gradle versions 

Author: ```suggestion
support the latest Gradle feature set. With this approach, it is possible to extend support for older Gradle versions 
```

with limited functionality.

Currently, there are only two variants of the Kotlin Gradle plugin:

* `main` for Gradle versions 6.7.1–6.9.2

> * `gradle70` for Gradle versions 7.0 and higher

Reviewer 2: It is more in Kotlin 1.8.0:
- `gradle71` for Gradle 7.1 - 7.4
- `gradle75` for Gradle 7.5
- `gradle76` for Gradle 7.6 and higher

In future Kotlin releases, more variants will be probably added.

To check which variant your build uses, enable
the [`--info` log level](https://docs.gradle.org/current/userguide/logging.html#sec:choosing_a_log_level) and find a
string in the output starting with `Using Kotlin Gradle plugin`, for example, `Using Kotlin Gradle plugin main variant`.

## Troubleshooting

> Here are workarounds for some known issues with variant selection in Gradle:
> * [ResolutionStrategy in pluginManagement is not working for plugins with multivariants](https://github.com/gradle/gradle/issues/20545)
> * [Plugin variants are ignored when a plugin is added as the `buildSrc` common dependency](https://github.com/gradle/gradle/issues/20847)
>
{type="note"}

### Gradle can't select a KGP variant in a custom configuration

This is an expected situation that Gradle can't select a KGP variant in a custom configuration.
If you use a custom Gradle configuration:

```kotlin
configurations.register("customConfiguraton") {
    ...
}
```

and want to add a dependency on the Kotlin Gradle plugin, for example:

```kotlin
dependencies {

>     compileOnly("org.jetbrains.kotlin:kotlin-gradle-plugin:%kotlinVersion%")

Reviewer 2: ```suggestion
    customConfiguration("org.jetbrains.kotlin:kotlin-gradle-plugin:%kotlinVersion%")
```

[//]: # (title: Support for Gradle plugin variants)

Gradle 7.0 introduced a new feature for Gradle plugin authors
— [plugins with variants](https://docs.gradle.org/7.0/userguide/implementing_gradle_plugins.html#plugin-with-variants).
This feature makes it easier to add support for latest Gradle features while maintaining compatibility with older Gradle versions

> below 7.1. Learn more about [variant selection in Gradle](https://docs.gradle.org/current/userguide/variant_model.html).

Reviewer 2: Maybe remove `below 7.1`?

With Gradle plugin variants, the Kotlin team can ship different Kotlin Gradle plugin (KGP) variants for different Gradle versions. 
The goal is to support the base Kotlin compilation in the `main` variant, which corresponds to the oldest supported versions of
Gradle. Each variant will have implementations for Gradle features from a corresponding release. The latest variant will
support the latest Gradle feature set. With this approach, it is possible to extend support for older Gradle versions 
with limited functionality.

Currently, there are only two variants of the Kotlin Gradle plugin:

> * `main` for Gradle versions 6.7.1–6.9.2

Reviewer 2: ```suggestion
* `main` for Gradle versions 6.8.3–6.9.2
```

Author: I suspected!😁

* `gradle70` for Gradle 7.0 and higher
* `gradle71` for Gradle 7.1-7.4
* `gradle75` for Gradle 7.5
* `gradle76` for Gradle 7.6 and higher

In future Kotlin releases, more variants will be probably added.

To check which variant your build uses, enable
the [`--info` log level](https://docs.gradle.org/current/userguide/logging.html#sec:choosing_a_log_level) and find a
string in the output starting with `Using Kotlin Gradle plugin`, for example, `Using Kotlin Gradle plugin main variant`.

## Troubleshooting

> Here are workarounds for some known issues with variant selection in Gradle:
> * [ResolutionStrategy in pluginManagement is not working for plugins with multivariants](https://github.com/gradle/gradle/issues/20545)
> * [Plugin variants are ignored when a plugin is added as the `buildSrc` common dependency](https://github.com/gradle/gradle/issues/20847)
>
{type="note"}

### Gradle can't select a KGP variant in a custom configuration

This is an expected situation that Gradle can't select a KGP variant in a custom configuration.
If you use a custom Gradle configuration:

```kotlin

> configurations.register("customConfiguraton") {

Author: @Tapchicoma, I'm sorry, could you please provide the Groovy code for this and the following two scripts?🥺

Reviewer 2: here Groovy will be the same

Reviewer 2: here Groovy will be the same

    ...
}
```

and want to add a dependency on the Kotlin Gradle plugin, for example:

```kotlin
dependencies {
    customConfiguration("org.jetbrains.kotlin:kotlin-gradle-plugin:%kotlinVersion%")

> }

Reviewer 2: Groovy:

```groovy
dependencies {
    customConfiguration 'org.jetbrains.kotlin:kotlin-gradle-plugin:%kotlinVersion%'
}
```

```

You need to add the following attributes to your `customConfiguration`:

```kotlin
configurations {
    customConfiguration {
        attributes {
            attribute(
                    Usage.USAGE_ATTRIBUTE,
                    project.objects.named(Usage.class, Usage.JAVA_RUNTIME)
            )
            attribute(
                    Category.CATEGORY_ATTRIBUTE,
                    project.objects.named(Category.class, Category.LIBRARY)
            )
            // If you want to depend on a specific KGP variant:
            attribute(
                 GradlePluginApiVersion.GRADLE_PLUGIN_API_VERSION_ATTRIBUTE,
                 project.objects.named("7.0")
            )
        }
    }

> }

Reviewer 2: Groovy:
```groovy
configurations {
    customConfiguration {
        attributes {
            attribute(
                    Usage.USAGE_ATTRIBUTE,
                    project.objects.named(Usage, Usage.JAVA_RUNTIME)
            )
            attribute(
                    Category.CATEGORY_ATTRIBUTE,
                    project.objects.named(Category, Category.LIBRARY)
            )
            // If you want to depend on a specific KGP variant:
            attribute(
                 GradlePluginApiVersion.GRADLE_PLUGIN_API_VERSION_ATTRIBUTE,
                 project.objects.named('7.0')
            )
        }
    }
}
```
